File: Stack/Leetcode/Hard/LargestRectangleInHistogram/LargestRectangleInHistogram.java
// https://leetcode.com/problems/largest-rectangle-in-histogram/
import java.util.ArrayDeque;
import java.util.Deque;

public class LargestRectangleInHistogram {

    public int largestRectangleArea(int[] heights) {
        // find nse, pse
        // for each index to rectangle will be height[i] * (nse[i] - pse[i] - 1)
        Deque<Integer> st = new ArrayDeque<>();
        int maxR = 0;
        for(int i=0; i<heights.length; i++) {
            while(!st.isEmpty() && heights[st.peek()] >= heights[i]) {
                int ind = st.pop();
                int left = st.isEmpty() ? -1 : st.peek();
                maxR = Math.max(maxR, heights[ind] * (i - left - 1));
            }
            st.push(i);
        }
        while(!st.isEmpty()) {
            int ind = st.pop();
            int left = st.isEmpty() ? -1 : st.peek();
            maxR = Math.max(maxR, heights[ind] * (heights.length - left - 1));
        }
        return maxR;
    }
}

File: Stack/Leetcode/Hard/LargestRectangleInHistogram/LargestRectangleInHistogram.cpp
// https://leetcode.com/problems/largest-rectangle-in-histogram/
#include<bits/stdc++.h>
using namespace std;
int largestRectangleArea(vector<int> &heights)
{
    stack<int> st;
    int n = heights.size();
    int maxArea = 0;
    for (int i = 0; i < n; i++)
    {
        while (!st.empty() && heights[st.top()] >= heights[i])
        {
            int ind = st.top();
            st.pop();
            int left = st.empty() ? -1 : st.top();
            maxArea = max(maxArea, heights[ind] * (i - left - 1));
        }
        st.push(i);
    }
    while (!st.empty())
    {
        int ind = st.top();
        st.pop();
        int left = st.empty() ? -1 : st.top();
        maxArea = max(maxArea, heights[ind] * (n - left - 1));
    }
    return maxArea;
}